anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.anr.internal.store
+
+import androidx.room.Database
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import androidx.room.RoomDatabase
+import androidx.room.migration.Migration
+
+@Database(
+    exportSchema = true,
+    version = 1,
+    entities = [
+        AnrInternalEntity::class,
+        CrashInternalEntity::class,
+    ],
+)
+abstract class CrashANRsInternalDatabase : RoomDatabase() {
+    abstract fun arnDao(): InternalANRDao
+    abstract fun crashDao(): InternalCrashDao
+
+    companion object {
+        val ALL_MIGRATIONS = emptyArray<Migration>()
+    }
+}
+
+@Entity(tableName = "anr_events")
+data class AnrInternalEntity(
+    @PrimaryKey val hash: String,
+    val message: String,
+    val name: String,
+    val file: String?,
+    val lineNumber: Int,
+    val stackTrace: String,
+    val timestamp: String,
+    val webView: String,
+    val customTab: Boolean,
+)
+
+@Entity(tableName = "crash_events")
+data class CrashInternalEntity(
+    @PrimaryKey val hash: String,
+    val shortName: String,
+    val processName: String,
+    val message: String,
+    val stackTrace: String,
+    val version: String,
+    val timestamp: String,
+    val webView: String,
+    val customTab: Boolean,
+)
```

Comment:
Why don't we use the same `AnrEntity` and `ExceptionEntity` we have alrady.

Attaching here a patch for reference

```diff
Index: anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/setting/CrashANRsListView.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/setting/CrashANRsListView.kt b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/setting/CrashANRsListView.kt
--- a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/setting/CrashANRsListView.kt	(revision edf018c8aa0f0b0b4227df26033801c1f24bbba5)
+++ b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/setting/CrashANRsListView.kt	(date 1718004720230)
@@ -74,13 +74,13 @@
         ViewTreeLifecycleOwner.get(this)?.lifecycleScope?.launch {
             crashANRsRepository.getANRs().combine(crashANRsRepository.getCrashes()) { anrs, crashes ->
                 (
-                    anrs.map {
+                    anrs.map { it.anr }.map {
                         AnrItem(
-                            stackTrace = it.stackTrace,
+                            stackTrace = it.stackTrace.joinToString("\n"),
                             customTab = it.customTab,
                             timestamp = it.timestamp,
                         )
-                    } + crashes.map {
+                    } + crashes.map { it.crash }.map {
                         CrashItem(
                             stackTrace = it.stackTrace,
                             processName = it.processName,
Index: anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashAnrsObserver.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashAnrsObserver.kt b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashAnrsObserver.kt
--- a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashAnrsObserver.kt	(revision edf018c8aa0f0b0b4227df26033801c1f24bbba5)
+++ b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashAnrsObserver.kt	(date 1718004880549)
@@ -35,7 +35,6 @@
 import kotlinx.coroutines.flow.flowOn
 import kotlinx.coroutines.launch
 import logcat.logcat
-import okio.ByteString.Companion.encode
 
 @ContributesMultibinding(
     scope = AppScope::class,
@@ -71,18 +70,10 @@
         logcat { "Observing ANRs database..." }
         anrDatabase.arnDao().getAnrsFlow().drop(1).distinctUntilChanged().flowOn(coroutineDispatcher.io()).collect { anrs ->
             anrs.forEach { anr ->
-                val stacktraceAsString = anr.stackTrace.joinToString("\n")
                 repository.insertANR(
                     AnrInternalEntity(
-                        hash = (stacktraceAsString + anr.timestamp).encode().md5().hex(),
-                        stackTrace = stacktraceAsString,
-                        timestamp = anr.timestamp,
-                        message = anr.message,
-                        name = anr.name,
-                        file = anr.file,
-                        lineNumber = anr.lineNumber,
-                        webView = anr.webView,
-                        customTab = anr.customTab,
+                        hashKey = anr.hash,
+                        anr = anr,
                     ),
                 )
             }
@@ -95,15 +86,8 @@
             crashes.forEach { crash ->
                 repository.insertCrash(
                     CrashInternalEntity(
-                        hash = (crash.stackTrace + crash.timestamp).encode().md5().hex(),
-                        stackTrace = crash.stackTrace,
-                        timestamp = crash.timestamp,
-                        shortName = crash.shortName,
-                        processName = crash.processName,
-                        message = crash.message,
-                        version = crash.version,
-                        webView = crash.webView,
-                        customTab = crash.customTab,
+                        hashKey = crash.hash,
+                        crash = crash,
                     ),
                 )
             }
Index: anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt
--- a/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt	(revision edf018c8aa0f0b0b4227df26033801c1f24bbba5)
+++ b/anrs/anrs-internal/src/main/java/com/duckduckgo/app/anr/internal/store/CrashANRsInternalDatabase.kt	(date 1718004880547)
@@ -17,10 +17,16 @@
 package com.duckduckgo.app.anr.internal.store
 
 import androidx.room.Database
+import androidx.room.Embedded
 import androidx.room.Entity
 import androidx.room.PrimaryKey
 import androidx.room.RoomDatabase
+import androidx.room.RoomWarnings
+import androidx.room.TypeConverters
 import androidx.room.migration.Migration
+import com.duckduckgo.app.anrs.store.AnrEntity
+import com.duckduckgo.app.anrs.store.AnrTypeConverter
+import com.duckduckgo.app.anrs.store.ExceptionEntity
 
 @Database(
     exportSchema = true,
@@ -30,6 +36,7 @@
         CrashInternalEntity::class,
     ],
 )
+@TypeConverters(AnrTypeConverter::class)
 abstract class CrashANRsInternalDatabase : RoomDatabase() {
     abstract fun arnDao(): InternalANRDao
     abstract fun crashDao(): InternalCrashDao
@@ -40,27 +47,15 @@
 }
 
 @Entity(tableName = "anr_events")
+@SuppressWarnings(RoomWarnings.PRIMARY_KEY_FROM_EMBEDDED_IS_DROPPED)
 data class AnrInternalEntity(
-    @PrimaryKey val hash: String,
-    val message: String,
-    val name: String,
-    val file: String?,
-    val lineNumber: Int,
-    val stackTrace: String,
-    val timestamp: String,
-    val webView: String,
-    val customTab: Boolean,
+    @PrimaryKey val hashKey: String,
+    @Embedded val anr: AnrEntity,
 )
 
 @Entity(tableName = "crash_events")
+@SuppressWarnings(RoomWarnings.PRIMARY_KEY_FROM_EMBEDDED_IS_DROPPED)
 data class CrashInternalEntity(
-    @PrimaryKey val hash: String,
-    val shortName: String,
-    val processName: String,
-    val message: String,
-    val stackTrace: String,
-    val version: String,
-    val timestamp: String,
-    val webView: String,
-    val customTab: Boolean,
+    @PrimaryKey val hashKey: String,
+    @Embedded val crash: ExceptionEntity,
 )
```

Reply:
I did consider it but I don't think we need to use the same entity and force the migrations. Actually, I still feel I prefer to have this decoupled from the real feature. I want to evaluate if this gives us the value I expect, and how we use it. In the future, I don't want to introduce a mandatory update in the internal feature if we add new things in the real db. I prefer to leave that part optional.